### PR TITLE
Add projection argument to Reference.fetch()

### PR DESCRIPTION
**Pull request by whophil**
Adds projection argument to `Reference.fetch()`, allowing partial reference documents to be returned.

Addresses _half_ of issue #196

### CODE REVIEW

File: umongo/frameworks/pymongo.py [hunk 1]
```diff
@@ -322,11 +322,14 @@ def __init__(self, *args, **kwargs):
...
-    def fetch(self, no_data=False, force_reload=False):
+    def fetch(self, no_data=False, force_reload=False, projection=None):
         if not self._document or force_reload:
             if self.pk is None:
                 raise NoneReferenceError('Cannot retrieve a None Reference')
-            self._document = self.document_cls.find_one(self.pk)
+            if projection is None:
+                self._document = self.document_cls.find_one(self.pk)
+            else:
+                self._document = self.document_cls.find_one(self.pk, projection)
```
Comment (lafrech) [inline umongo/frameworks/pymongo.py:332]: Why not a single case

    self._document = self.document_cls.find_one(self.pk, projection=projection)

Doesn't this work for all cases?

**Review by lafrech — COMMENTED** [review]
I left a minor comment.

The issue you point to points to #98 which raises concerns about projection and caching mechanism. And the fact that DB names should be expected (unless umongo tries to translates field names into DB names but I think it wouldn't be the only place where DB names are needed).

In any case, you may merge this if you're happy with it. If we can't have a perfect solution, this is better than nothing. It would be nice if known limitations were documented, though.

**Comment by whophil** [comment]
Thanks @lafrech. I had actually not seen the discussion on issue #98. Let me think over this a bit and potentially rework.

**Comment by whophil** [comment]
@lafrech I've implemented `cook_find_projection`, as suggested in #98, so that the projections can use uMongo field names and not DB names.

Other things discussed in #98 which I did not address:
- The cached `_document` is just the first-fetched document, not a cache which is looked up by the projection. The projection-keyed cache may still be useful in the future. If implemented, it should have a configurable max size.
- `_reference_io_validate` is untouched (it still uses `.fetch()`).

**Comment by whophil** [comment]
FYI, the suggestion to not use`.fetch()` in `_reference_io_validate` makes a huge difference in performance in some cases.

In my application, I have a document which references over a thousand others. The time to `.commit()` the parent document went from 21.8 seconds to 1.5 seconds (PyMongo with mongo running on localhost).

File: umongo/data_objects.py [hunk 2]
```diff
@@ -165,19 +165,28 @@ def __init__(self, document_cls, pk):
...
+        :param no_data: if True, the caller is only interested in whether
+            the document is present in database. This means the
             implementation may not retrieve document's data to save bandwidth.
         :param force_reload: if True, ignore any cached data and reload referenced
             document from database.
+        :param projection: if supplied, this is a dictionary or list describing
+            a projection which limits the data returned from database.
         """
         raise NotImplementedError
     # TODO replace no_data by `exists` function
```
Comment (lafrech) [inline umongo/data_objects.py:181]: I guess we could deprecate `no_data`, then? (No pressure to do this in this PR.)

**Review by lafrech — APPROVED** [review]
LGTM!

You may merge and release!

File: umongo/data_objects.py (hunk 2 above)
Comment (whophil) [reply umongo/data_objects.py]: Sounds good, this can be deprecated in a future PR/release.

**Comment by whophil** [comment]
@lafrech how do I make the CI "go" on this?

**Comment by lafrech** [comment]
That's a good question. I see no obvious reason. Hooks are enabled. Can't look into this right now.

Frankly, Azure CI is a nightmare to configure and rather than trying to fix this I'd be tempted to move to GitHub actions. You may give it a try in another PR if you want. Setting up tests is relatively straightforward. See what I did on flask-smorest. The tricky part would be launching MongoDB.

------------------------------

OK I peeked into Azure logs and CI did go but apparently didn't comment here with the results. Note sure if the pipeline logs there are public.

Here's the last result:

```
./tests/conftest.py:38:101: E501 line too long (106 > 100 characters)
./umongo/frameworks/motor_asyncio.py:415:101: E501 line too long (105 > 100 characters)
```

**Comment by whophil** [comment]
> OK I peeked into Azure logs and CI did go but apparently didn't comment here with the results. Note sure if the pipeline logs there are public.

I don't know where to find the link to these results, so I can't tell you if they're public or not :)

I fixed the too-long lines. Merging now!

As far as a release goes - what is the process on this project?

**Pull request merged**